[service-loadbalancer] SSL termination support with HAProxy

remoe opened the pull request:
This pull request adds support for SSL termination.
## Installation instructions
- Add annotations to your service:

``` yaml
metadata:
  name: myservice
  annotations:
    serviceloadbalancer/lb.sslTerm: "true"
    serviceloadbalancer/lb.aclMatch: "-i /t1 /t2"
  labels:
```
- Create a secret with one of your favorite tool
- Update your loadbalancer POD to something like the following

``` yaml
        ports:
        # All http services
        - containerPort: 80
          hostPort: 80
          protocol: TCP
        # ssl term
        - containerPort: 443
          hostPort: 443
          protocol: TCP       
        # haproxy stats
        - containerPort: 1936
          hostPort: 1936
          protocol: TCP
        resources: {}     
        volumes:
        - name: secret-volume
          secret:
            secretName: my-secret
        volumeMounts:
        - mountPath: "/ssl"
          name: secret-volume
```
- Add your SSL configuration to loadbalancer POD

``` yaml
      args:
      - --ssl-cert=/ssl/crt.pem
      - --ssl-ca-cert=/ssl/ca.crt
      - --namespace=default
```
- Define your https-frontend section in template.cfg

```
{{ if ne .sslCert "" }}
frontend httpsfrontend
    mode http
    bind :443 ssl {{ .sslCert }} ciphers ALL:!HIGH:!LOW:!MD5:!SEED:!EXP:MEDIUM:RC4-SHA verify required

{{range $i, $svc := .services.httpsTerm}}
    acl url_acl_{{$svc.Name}} path_beg {{$svc.AclMatch}}
    {{ if $svc.Host }}acl host_acl_{{$svc.Name}} hdr(host) {{$svc.Host}}
    use_backend {{$svc.Name}} if url_acl_{{$svc.Name}} or host_acl_{{$svc.Name}}
    {{ else }}use_backend {{$svc.Name}} if url_acl_{{$svc.Name}}
{{ end }}
{{end}}
{{end}}
```

Im not that experienced in english to be able to update the readme for instructions. I hope someone can help to finish this up.

remoe committed:
SSL termination support with HAProxy

bprashanth commented:
NIce! thanks, I'll get to this sometime soon. In the meanwhile @aledbf might have some review comments.

bprashanth commented:
> copy your cert and ca for example to: /cfg/ssl/

@remoe this is fine for a first step, but if you're interested in taking this a little further I suggest storing the certs in a secret that is mentioned in the annotation. The loadbalancer would have to read the secrets and load write files. 

Simple code to write out pems from key/cert: https://github.com/bprashanth/Ingress/blob/master/examples/https/haproxy/daemonizer.go#L109

And readin secrets should also be easy:
```
client.Secrets(namespace).Get(name)
```

You can throw an error if the secret is not a key/cert combo, write out the pem, test the new config (I think haproxy -f) and avoid reloading if the secrets are, say, not in the correct format.

remoe commented:
@bprashanth thanks for your hint. I've updated the tests.
Travis fails at "verify-gofmt.sh". But I don't see any output of https://github.com/kubernetes/contrib/blob/master/hack/verify-gofmt.sh#L50
Do you know what is wrong with it?

aledbf commented:
@remoe you can check the error running `./hack/verify-gofmt.sh` from the root of the contrib git directory

aledbf commented:
@remoe can you make the `httpsfrontend` conditional?

bprashanth commented:
>Add annotations to your service:

Just fyi you can also annotate at runtime via kubectl annotate.

>Copy your cert and ca for example to: /cfg/ssl/

You should make this a secret, even if loading them isn't dynamic we shouldn't require users to copy and mount. The kubelet will do that for you. Here's an example of creating/defining a secret in pod spec: https://github.com/kubernetes/kubernetes/blob/master/examples/https-nginx/nginx-app.yaml#L42

Can we actually just get away with duplicating the instruction here to make secret? https://github.com/kubernetes/kubernetes/blob/master/examples/https-nginx/README.md#generate-certificates

> Define your https-frontend section in template.cfg

As Manuel suggested we should make this optional

remoe commented:
@bprashanth @aledbf , about filtering: do you mean something like this:

```golang
{{ if .sslCrt }}
frontend httpsfrontend
...
{{end}}
```

remoe commented:
I found that sslCrt, sslCa and Ciphers are not needed. I will only add sslCert as a pem file, because of:

https://cbonte.github.io/haproxy-dconv/configuration-1.6.html#5.1-crt

```
If your CA
requires an intermediate certificate, this can also be concatenated into this
file.
```

And the ciphers can be manually inserted in the template.

remoe commented:
@bprashanth when i understand you correctly then you mean one don't need to define any volume on the POD (also no secret volume). One only define a secret name and the load-balancer get the files from this secret. Should this files explicit defined or with a console argument?
But then the load-balancer must write temporary files for HAProxy, right?

bprashanth commented:
Sorry for the misunderstanding. There are 2 scenarios:
1. ssl termination with lb certs: this needs a secret + secret volumen definition in pod
2. ssl termination with service certs: this needs an annotation on the service and the loadbalancer needs to read the secrets from the apiserver and write them out to files.

remoe commented:
@aledbf I've make httpsfrontend conditional. Do you know why travis not allow my code? I've changed known-flags.txt . Do I need also change the mungegithub/whitelist.txt ?

@bprashanth I've implemented scenario 1. I haven't time for 2, sorry.

bprashanth commented on service-loadbalancer/service_loadbalancer.go:
```diff
@@ -277,6 +301,13 @@ func (cfg *loadBalancerConfig) write(services map[string][]service, dryRun bool)
+		conf["sslCert"] = "crt " + cfg.sslCert
+	}
+	if cfg.sslCaCert != "" {
+		conf["sslCert"] = conf["sslCert"].(string) + " ca-file " + cfg.sslCaCert
```
can you avoid the assertion by constructing the string first and then just assigning it to sslCert?

bprashanth replied:
Actually, since you're getting it at loadbalancer bootstrap time, it will never change so there's no need to recreate it on each reload. How about just storing it as cfg.sslConfig, and writing a comment with an example of how it might look?

remoe replied:
But then one need a substitution in the config "sslConfig" like:

```golang
crt {{ .sslCert }} ca-file {{ .sslCaCert}} ciphers
```

remoe replied:
I've tested this on my kubernetes cluster and it works. So I think it is ready to merge for a first version. Its a little bit painful to develop for kubernetes cluster on a windows system. So it takes a lot of time for me, sorry.

bprashanth replied:
Yeah i leave the second part to you, either is fine. please do this though:
```go
if cfg.sslCert != "" {
  sslConfig = "crt "+cfg.sslCert
}
if cfg.sslCaCert {
  sslCert += "ca-file  ..."
}
conf["sslCert"] = sslCert
```

bprashanth commented on service-loadbalancer/service_loadbalancer.go:
```diff
@@ -152,6 +158,12 @@ type service struct {
 	// host header inside the http request. It only applies to http traffic.
 	Host string
 
+	// if true it apply to ssl termination
```
please modify to: if true, terminate ssl using the loadbalancers certificates.

bprashanth commented:
go code LGTM but for a few nits, I'm assuming the new template works as intended because I don't know the haproxy cfg format off the top of my head. @aledbf might want to take a final look at the haproxy config changes.

bprashanth commented:
Please also update the README with a section like:
```
## SSL Termination
1. Add secrets to your loadbalancer pod
2. Annotate a service
```

That's it right?

aledbf commented:
@remoe can you provide better defaults? Something like:
```
in global section
{{ if ne .sslCert "" }}
    ssl-default-bind-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!3DES:!MD5:!PSK
    ssl-default-bind-options no-tls-tickets
{{ end }}

{{ if ne .sslCert "" }}
frontend httpsfrontend
    mode http
    bind :443 ssl {{ .sslCert }} no-sslv3

    # HSTS (15768000 seconds = 6 months)
    rspadd  Strict-Transport-Security:\ max-age=15768000

{{range $i, $svc := .services.httpsTerm}}
    acl url_acl_{{$svc.Name}} path_beg {{$svc.AclMatch}}
    {{ if $svc.Host }}acl host_acl_{{$svc.Name}} hdr(host) {{$svc.Host}}
    use_backend {{$svc.Name}} if url_acl_{{$svc.Name}} or host_acl_{{$svc.Name}}
    {{ else }}use_backend {{$svc.Name}} if url_acl_{{$svc.Name}}
{{ end }}
{{end}}
{{end}}
```

[Here]( https://mozilla.github.io/server-side-tls/ssl-config-generator/) you can see different examples

remoe commented:
@bprashanth @aledbf done.

remoe commented:
@bprashanth anything else I can do?

bprashanth commented:
@remoe didn't notice you had uploaded a new patch set. Please squash your commits and rebase, and I'll merge. Also please run  hack/verify-flags-underscore.py locally and see why travis is complaining. Thanks.

remoe commented:
@bprashanth about verify-flags-underscore.py, I see the following:

```shell
../hack/verify-flags-underscore.py
Found illegal 'flag' usage. If these are false positives you should run `hack/verify-flags-underscore.py -e > hack/verify-flags/exceptions.txt` to update the list.
Ingress/controllers/nginx-third-party/nginx.tmpl:        require("error_page")
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 403 @custom_403;
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 404 @custom_404;
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 405 @custom_405;
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 408 @custom_408;
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 413 @custom_413;
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 500 @custom_500;
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 501 @custom_501;
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 502 @custom_502;
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 503 @custom_503;
Ingress/controllers/nginx-third-party/nginx.tmpl:    error_page 504 @custom_504;
```
not in my changes ?!

bprashanth commented:
@remoe Hmm, that would mean HEAD is failing. Can you try on your master branch without your changes? and if so, please open a bug or send a pr.

remoe commented:
@bprashanth about squash/rebase: Is this now ok?

bprashanth commented:
you actually seem to have 14 commits (or github is confused). We typically take 1 or 2 well defined commits so they're easier to revert. Fyi I generally squash with something like `git rebase -i HEAD~14`.

remoe commented:
@bprashanth sorry. I hope this works now. I don't know how to solve this verify-flags-underscore issue.

remoe commented:
@bprashanth sorry. I hope this works now. I don't know how to solve this verify-flags-underscore issue.

bprashanth commented:
ok, this is good. just do as verify-flags-underscore indicates (with the -e option) and I'll file a bug

remoe committed:
update verify-flags exceptions

remoe commented:
@bprashanth I've updated the verify-flags exceptions. Is it now ready?

bprashanth commented:
@remoe thanks

remoe commented:
@bprashanth @aledbf thanks for your help.
